src/panel-toc-editor.ts

```diff
@@ -0,0 +1,259 @@
+  };
+}
+
+function moduleToObject(element: any): any {
+	const moduleid = element.getAttribute('document');
+	if (moduleid == null) {
+		throw new Error('Error parsing collection. ModuleID missing.');
+	}
+  return moduleObjectFromModuleId(moduleid);
+}
+
+function subcollectionToObject(element: any): any {
+  const title = element.getElementsByTagNameNS(NS_METADATA, 'title')[0].textContent;
+	const content = element.getElementsByTagNameNS(NS_COLLECTION, 'content')[0];
+	if (title == null) {
+		throw new Error('Error parsing collection. Subcollection title missing.');
+	}
+  return {
+    type: 'subcollection',
+    title: title,
+    children: childObjects(content)
+  };
+}
+
+function childObjects(element: any): Array<any> {
+	const children = [];
+  for (const child of Array.from<any>(element.childNodes)) {
+    if (child.localName === 'module') {
+      children.push(moduleToObject(child));
+    } else if (child.localName === 'subcollection') {
+      children.push(subcollectionToObject(child));
+    }
+  }
+  return children;
+}
+
+function parseCollection(xml: string): any {
+	const document = new DOMParser().parseFromString(xml);
+
+  const metadata = document.getElementsByTagNameNS(NS_COLLECTION, 'metadata')[0];
+  const collectionTitle = metadata.getElementsByTagNameNS(NS_METADATA, 'title')[0].textContent;
+  const collectionSlug = metadata.getElementsByTagNameNS(NS_METADATA, 'slug')[0].textContent;
+  
+  const treeRoot = document.getElementsByTagNameNS(NS_COLLECTION, 'content')[0];
+
+  const tree = {
+    type: 'collection',
+		title: collectionTitle,
+		slug: collectionSlug,
+    children: childObjects(treeRoot)
+  };
+
+  return tree;
+}
+
+function populateTreeDataToXML(document: XMLDocument, root: any, treeData: any) {
+	for (const child of treeData.children) {
+		const element = document.createElementNS(NS_COLLECTION, child.type);
+		 // md prefix is technically a guess. If incorrect, document may have a lot of xmlns:md attributes
+		const title = document.createElementNS(NS_METADATA, 'md:title');
+		const titleContent = document.createTextNode(child.title);
+		title.appendChild(titleContent);
+		element.appendChild(title);
+		root.appendChild(element);
+		if (child.type === 'subcollection') {
+			const contentWrapper = document.createElementNS(NS_COLLECTION, 'content');
+			element.appendChild(contentWrapper);
+			populateTreeDataToXML(document, contentWrapper, child);
+		} else if (child.type === 'module') {
+			element.setAttribute('document', child.moduleid);
+		}
+	}
+}
+
+function replaceCollectionContent(document: XMLDocument, treeData: any) {
+	const content = document.getElementsByTagNameNS(NS_COLLECTION, 'content')[0];
+
+	const newContent = document.createElementNS(NS_COLLECTION, 'content');
+	content.parentNode!.replaceChild(newContent, content);
+	populateTreeDataToXML(document, newContent, treeData);
+}
+
+
+function getModuleTitle(moduleid: string): string {
+	if (guessedModuleTitles.hasOwnProperty(moduleid)) {
```

Comment:
It is not discarded per say, but the `guessedModuleTitles` object is updated with new values every time the ToC Editor panel is opened, since `preloadGuessedModuleTitles` is called upon the command being run. One thing I'd like to get done in the future is the titles in the panel staying up to date without requiring a close/open or refresh of some kind.